### PR TITLE
Modifications to run custom_files/run_validation.test.sh on CPU

### PR DESCRIPTION
**Pull request by LoreggiaGiacomo**

### CODE REVIEW

File: scene_graph_benchmark/relation_head/sparse_targets.py [hunk 1]
```diff
@@ -79,8 +80,10 @@ def _get_tensor_from_boxlist(proposals, field='labels'):
             im_inds = torch.cat(
                 (im_inds, im_ind * torch.ones(num_proposals_im, 1)), dim=0)
 
-    # TODO: support both cpu and gpu
-    im_inds_batch = torch.Tensor(im_inds).long().cuda()
+    im_inds_batch = torch.Tensor(im_inds).long()
+    if framework_device == "cuda":
+        im_inds_batch.to(torch.device("cuda"))
```
Comment (kuzeko) [inline scene_graph_benchmark/relation_head/sparse_targets.py]: are you sure this is how to use the `to()` method?
My understanding is that you have to write:

```
im_inds_batch = im_inds_batch.to(torch.device("cuda"))
```

File: tools/demo/demo_image.py [hunk 2]
```diff
@@ -60,15 +60,19 @@ def main():
...
     parser.add_argument("opts", default=None, nargs=argparse.REMAINDER,
                         help="Modify config options using the command-line")
 
-    if not torch.cuda.is_available():
-        cfg.MODEL.DEVICE = "cpu"
-
     args = parser.parse_args()
+
+    if not torch.cuda.is_available() and args.device=="cuda":
+        raise RuntimeError("No GPU available. Please check the device selected or set up again the software correctly")
```
Comment (kuzeko) [inline tools/demo/demo_image.py]: error message can be more explicit. what is "set up again correctly" ?

**Review by kuzeko — CHANGES_REQUESTED** [review]
Check if the usage of the `to()` method is correct and maybe rephrase the error message.

See inline comments

File: scene_graph_benchmark/relation_head/sparse_targets.py (hunk 1 above)
Comment (LoreggiaGiacomo) [reply scene_graph_benchmark/relation_head/sparse_targets.py]: It is not necessary. From the documentation example (https://pytorch.org/docs/stable/generated/torch.Tensor.to.html):

```
tensor = torch.randn(2, 2)  # Initially dtype=float32, device=cpu
tensor.to(torch.float64)
tensor([[-0.5044,  0.0005],
        [ 0.3310, -0.0584]], dtype=torch.float64)

cuda0 = torch.device('cuda:0')
tensor.to(cuda0)
tensor([[-0.5044,  0.0005],
        [ 0.3310, -0.0584]], device='cuda:0')
```

Anyway, I tried on the software to be sure. Here is the modification I have made on the `sparse_targets.py` file:

image

and here is the output:

image

File: tools/demo/demo_image.py (hunk 2 above)
Comment (LoreggiaGiacomo) [reply tools/demo/demo_image.py]: Right, I will need to modify the README file as well to account for the CPU case.

File: scene_graph_benchmark/relation_head/sparse_targets.py (hunk 1 above)
Comment (kuzeko) [reply scene_graph_benchmark/relation_head/sparse_targets.py]: The example is misleading. because the print out is the print of the return value.
You test needa to be run on a machine with both gpu and cpu and you should do:
```
tensor_original = torch.randn(2, 2)  # Initially dtype=float32, device=cpu

cuda0 = torch.device('cuda:0')


tensor_returned = tensor_original.to(cuda0)

print (tensor_original)
print (tensor_returned)


```

Comment (LoreggiaGiacomo) [reply scene_graph_benchmark/relation_head/sparse_targets.py]: You are right, I am correcting it.
